packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol

```diff
@@ -345,17 +357,32 @@ abstract contract CrossDomainMessenger is
             "CrossDomainMessenger: message has already been relayed"
         );
 
+        // If there is not enough gas left to perform the external call and finish the execution,
+        // return early and assign the message to the failedMessages mapping.
         // If `xDomainMsgSender` is not the default L2 sender, this function
         // is being re-entered. This marks the message as failed to allow it
         // to be replayed.
-        if (xDomainMsgSender != Constants.DEFAULT_L2_SENDER) {
+        if (
+            gasleft() < _minGasLimit + RELAY_RESERVED_GAS ||
+            xDomainMsgSender != Constants.DEFAULT_L2_SENDER
+        ) {
             failedMessages[versionedHash] = true;
             emit FailedRelayedMessage(versionedHash);
+
+            // Revert in this case if the transaction was triggered by the estimation address. This
+            // should only be possible during gas estimation or we have bigger problems. Reverting
+            // here will make the behavior of gas estimation change such that the gas limit
+            // computed will be the amount required to relay the message, even if that amount is
+            // greater than the minimum gas limit specified by the user.
+            if (tx.origin == Constants.ESTIMATION_ADDRESS) {
+                revert("CrossDomainMessenger: failed to relay message");
+            }
+
             return;
         }
 
         xDomainMsgSender = _sender;
-        bool success = SafeCall.callWithMinGas(_target, _minGasLimit, _value, _message);
+        bool success = SafeCall.call(_target, gasleft(), _value, _message);
```

Comment:
This doesn't actually reserve the gas amount

Reply:
Still passing in the full gas amount, the gas buffer needs to be reserved explicitly (gasleft() - relay gas buffer). I think best to just use callWithMinGas here and add a new parameter for reservedGas. It's much clearer.

Reply:
The reason I didn't fully reserve the remaining gas was because the `RELAY_RESERVED_GAS` constant factors in both the worst case cost of the `CALL` opcode (minus the memory expansion / execution costs) as well as the gas required to finish the execution. We could split it up into 2 constants?

fwiw, I initially went w/ modifying `callWithMinGas`, but @maurelian brought up a great point that we can just not revert in this context if a threshold of remaining gas is not met, which is much safer. See the thread above in this PR ^

Reply:
> See the thread above in this PR ^

@smartcontracts the thread is gone from this view, please take a look at it [here](https://github.com/ethereum-optimism/optimism/pull/5470#discussion_r1169217738).

Reply:
> We could split it up into 2 constants?

I believe it was that way previously and that it makes the most sense, at least to my mind.

Reply:
Replacing `callWithMinGas` with `call` just has the effect of not explicitly guaranteeing that the sub-call will receive enough gas. So the minimum gas limit is no longer respected. I feel like this is trying to make a larger diff than the original intention which should be like a 10 line diff (add two buffers, make sure it always respects minimum gas limit).

I do understand that minimum gas limit is a pain here and we should consider modifying things in the future but I feel like this is a large unnecessary diff at this point

Reply:
> Replacing `callWithMinGas` with `call` just has the effect of not explicitly guaranteeing that the sub-call will receive enough gas.

It's true that `callWithMinGas` explicitly guarantees that the sub-call will receive enough gas, but it enforces that guarantee by bricking the withdrawal.

Reply:
Resolving on the basis that this applies to outdated code.